toolkit/offchain/src/register.rs

```diff
@@ -130,7 +130,94 @@ pub async fn run_register<
 	log::info!("✅ Transaction submitted. ID: {}", hex::encode(result.transaction.id));
 	await_tx.await_tx_output(ogmios_client, UtxoId::new(tx_id, 0)).await?;
 
-	Ok(Some(result.transaction))
+	Ok(Some(McTxHash(result.transaction.id)))
+}
+
+pub trait Deregister {
+	#[allow(async_fn_in_trait)]
+	async fn deregister(
+		&self,
+		genesis_utxo: UtxoId,
+		payment_signing_key: MainchainPrivateKey,
+		own_pkh: MainchainAddressHash,
+		stake_ownership_pub_key: MainchainPublicKey,
+	) -> Result<Option<McTxHash>, OffchainError>;
+}
+
+impl<T> Deregister for T
+where
+	T: QueryLedgerState + Transactions + QueryNetwork + QueryUtxoByUtxoId,
+{
+	async fn deregister(
+		&self,
+		genesis_utxo: UtxoId,
+		payment_signing_key: MainchainPrivateKey,
+		own_pkh: MainchainAddressHash,
+		stake_ownership_pub_key: MainchainPublicKey,
+	) -> Result<Option<McTxHash>, OffchainError> {
+		run_deregister(
+			genesis_utxo,
+			payment_signing_key,
+			own_pkh,
+			stake_ownership_pub_key,
+			self,
+			FixedDelayRetries::two_minutes(),
+		)
+		.await
+		.map_err(|e| OffchainError::InternalError(e.to_string()))
+	}
+}
+
+pub async fn run_deregister<
+	C: QueryLedgerState + QueryNetwork + QueryUtxoByUtxoId + Transactions,
+	A: AwaitTx,
+>(
+	genesis_utxo: UtxoId,
+	payment_signing_key: MainchainPrivateKey,
+	own_pkh: MainchainAddressHash,
+	stake_ownership_pub_key: MainchainPublicKey,
+	ogmios_client: &C,
+	await_tx: A,
+) -> anyhow::Result<Option<McTxHash>> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key.0, ogmios_client).await?;
+	let validator = crate::scripts_data::registered_candidates_scripts(genesis_utxo)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let all_registration_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
+	let own_registrations =
+		get_own_registrations(own_pkh, stake_ownership_pub_key.clone(), &all_registration_utxos);
+
+	if own_registrations.is_empty() {
+		log::info!("✅ Candidate is not registered.");
+		return Ok(None);
+	}
+
+	let own_registration_utxos = own_registrations.iter().map(|r| r.0.clone()).collect::<Vec<_>>();
```

Comment:
This could be moved to `get_own_registrations`

Reply:
`get_own_registrations` return `(OgmiosUtxo, CandidateRegistration)` pairs the second component of which is used in `register`